File: backend/app/models/analysis.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, Float, Text, DateTime, ForeignKey, JSON
from sqlalchemy.orm import relationship

from app.database import Base


class Analysis(Base):
    """AI-powered analysis of why a viral post went viral."""

    __tablename__ = "analyses"

    id = Column(Integer, primary_key=True)
    viral_post_id = Column(Integer, ForeignKey("viral_posts.id", ondelete="CASCADE"), unique=True, nullable=False)

    # AI-generated summary
    why_viral_summary = Column(Text, nullable=True)

    # Algorithm factors (0-100 scores)
    posting_time_score = Column(Float, nullable=True)
    hook_strength_score = Column(Float, nullable=True)
    engagement_velocity_score = Column(Float, nullable=True)
    save_share_ratio_score = Column(Float, nullable=True)
    hashtag_performance_score = Column(JSON, nullable=True)
    audience_retention_score = Column(Float, nullable=True)

    # Qualitative factors
    emotional_trigger = Column(String, nullable=True)  # joy|awe|anger|surprise|sadness|fear

    # AI confidence
    confidence_score = Column(Float, nullable=True)  # 0.0-1.0

    # Phase 5 fields: Audience insights
    audience_demographics = Column(JSON, nullable=True)
    # Structure: {
    #   "age_range": {"13-17": 5, "18-24": 25, "25-34": 45, "35-44": 20, "45+": 5},
    #   "gender_distribution": {"male": 40, "female": 55, "other": 5},
    #   "top_countries": [
    #     {"code": "US", "percentage": 45},
    #     {"code": "GB", "percentage": 12},
    #     {"code": "CA", "percentage": 8}
    #   ]
    # }

    engagement_rate = Column(Float, nullable=True)  # Percentage (0-100)
    # Calculation: (likes + comments + saves + shares) / follower_count * 100

    audience_interests = Column(JSON, nullable=True)
    # Structure: {
    #   "inferred_topics": ["fitness", "nutrition", "wellness"],
    #   "content_affinity": ["educational", "inspirational", "lifestyle"],
    #   "hashtag_analysis": ["#fitnessmotivation", "#healthylifestyle"]
    # }

    content_category = Column(String, nullable=True)
    niche = Column(String, nullable=True)

    # Phase 05-08: User niche override
    user_niche_override = Column(String, nullable=True)
    # When user provides override, store here. Overrides AI-detected niche field for display.

    created_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    viral_post = relationship("ViralPost", back_populates="analysis")

    def __repr__(self):
        return f"<Analysis(id={self.id}, viral_post_id={self.viral_post_id})>"

File: frontend/src/types/analysis.ts
export interface AudienceDemographics {
  age_range?: {
    [key: string]: number;  // e.g., "18-24": 25
  };
  gender_distribution?: {
    [key: string]: number;  // e.g., "male": 40
  };
  top_countries?: Array<{
    code: string;  // ISO country code
    percentage: number;
  }>;
}

export interface AudienceInterests {
  inferred_topics?: string[];  // e.g., ["fitness", "wellness"]
  content_affinity?: string[];  // e.g., ["educational", "inspirational"]
  hashtag_analysis?: string[];
  inferred_formats?: string[];  // Extended content formats from Phase 05-03
  categorization_confidence?: number;
  categorization_reason?: string;
  niche?: string;  // Primary niche from Phase 05-05
  niche_secondary?: string | null;
  niche_confidence?: number;
  niche_reasoning?: string;
  niche_keywords?: string[];
}

export interface Analysis {
  id: number;
  viral_post_id: number;

  // Phase 04: OpenAI Analysis
  why_viral_summary: string | null;
  posting_time_score: number | null;
  hook_strength_score: number | null;
  emotional_trigger: string | null;
  engagement_velocity_score: number | null;
  save_share_ratio_score: number | null;
  hashtag_performance_score: number | null;
  audience_retention_score: number | null;
  confidence_score: number | null;

  // Phase 05: Enriched Data
  engagement_rate?: number | null;  // Percentage (0-100+)
  content_category?: string | null;  // Instagram native type (Reel, Post, etc.)
  niche?: string | null;  // Detected niche
  audience_demographics?: AudienceDemographics | null;
  audience_interests?: AudienceInterests | null;
  user_niche_override?: string | null;  // User-customized niche (overrides AI-detected niche)

  created_at: string;
  updated_at?: string;
}
